core/comment.js

```diff
@@ -177,71 +208,99 @@ Blockly.Comment.prototype.resizeBubble_ = function() {
-    // Dispose of the bubble.
-    this.bubble_.dispose();
-    this.bubble_ = null;
-    this.textarea_ = null;
-    this.foreignObject_ = null;
+    this.disposeBubble_();
   }
-  // Restore the bubble stats after the visibility switch.
-  this.setText(text);
-  this.setBubbleSize(size.width, size.height);
 };
 
 /**
- * Bring the comment to the top of the stack when clicked on.
+ * Show the bubble. Handles deciding if it should be editable or not.
+ * @private
+ */
+Blockly.Comment.prototype.createBubble_ = function() {
+  if (!this.block_.isEditable() || Blockly.utils.userAgent.IE) {
+    this.createNonEditableBubble_();
+  } else {
+    this.createEditableBubble_();
+  }
+};
+
+/**
+ * Show an editable bubble.
+ * @private
+ */
+Blockly.Comment.prototype.createEditableBubble_ = function() {
+  this.bubble_ = new Blockly.Bubble(
+      /** @type {!Blockly.WorkspaceSvg} */ (this.block_.workspace),
+      this.createEditor_(), this.block_.svgPath_,
+      this.iconXY_, this.model_.size.width, this.model_.size.height);
+  // Expose this comment's block's ID on its top-level SVG group.
+  this.bubble_.setSvgId(this.block_.id);
+  this.bubble_.registerResizeEvent(this.onBubbleResize_.bind(this));
+  this.updateColour();
+};
+
+/**
+ * Show a non-editable bubble.
+ * @private
+ */
+Blockly.Comment.prototype.createNonEditableBubble_ = function() {
+  // Steal the code from warnings to make an uneditable text bubble.
+  // MSIE does not support foreignobject; textareas are impossible.
+  // https://docs.microsoft.com/en-us/openspecs/ie_standards/ms-svg/56e6e04c-7c8c-44dd-8100-bd745ee42034
+  // Always treat comments in IE as uneditable.
+  // TODO (#2917): It would be great if the comment could support line breaks.
+  Blockly.Warning.prototype.createBubble.call(this, true);
+};
+
+/**
+ * Dispose of the bubble.
+ * @private
+ */
+Blockly.Comment.prototype.disposeBubble_ = function() {
+  if (this.paragraphElement_) {
+    // We're using the warning UI so we have to let it dispose.
+    Blockly.Warning.prototype.disposeBubble.call(this);
+    return;
+  }
+
+  this.bubble_.dispose();
+  this.bubble_ = null;
+  this.textarea_ = null;
+  this.foreignObject_ = null;
+};
+
+/**
+ * Callback fired when an edit starts.
+ *
+ * Bring the comment to the top of the stack when clicked on. Also cache the
+ * current text so it can be used to fire a change event.
  * @param {!Event} _e Mouse up event.
  * @private
  */
-Blockly.Comment.prototype.textareaFocus_ = function(_e) {
-  // Ideally this would be hooked to the focus event for the comment.
-  // However doing so in Firefox swallows the cursor for unknown reasons.
-  // So this is hooked to mouseup instead.  No big deal.
+Blockly.Comment.prototype.startEdit_ = function(_e) {
   if (this.bubble_.promote_()) {
     // Since the act of moving this node within the DOM causes a loss of focus,
     // we need to reapply the focus.
     this.textarea_.focus();
   }
+
+  this.cachedText_ = this.model_.text;
```

Comment:
type mismatch here. this.model_.text is string or null. cachedText is expecting a string.